### PR TITLE
Implement staff assignment for Caseload

### PR DESCRIPTION
**Pull request by martha**
[//]: # 'remove this if PR is for a release-* branch'
# _Please squash merge this PR_

## Description
Ticket: https://github.com/open-path/Green-River/issues/6335
Spec doc: https://docs.google.com/document/d/17o9wMCrYR1Wo4vX5iLUqYMe1b7pZQjDdhFCQx6_T5-M/edit

- Adds a migration to create the tables indicated in the spec doc
- Adds new models and schema objects for the new types
- Adds a new project config type for indicating whether projects have staff assignment enabled
- Adds assign/unassign mutations

To test: It's probably easiest to test this in conjunction with the frontend changes, which are still a draft, but I've pushed what I have so far here: https://github.com/greenriver/hmis-frontend/pull/856

## Type of change
[//]: # 'remove options that are not relevant'
- [ ] Bug fix
- [x] New feature (adds functionality)
- [ ] Code clean-up / housekeeping
- [ ] Dependency update

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

### CODE REVIEW

File: drivers/hmis/app/graphql/types/forms/pick_list_option.rb
Comment (martha) [inline drivers/hmis/app/graphql/types/forms/pick_list_option.rb]: to discuss: Is this permission check correct / is it overkill?

File: drivers/hmis/app/graphql/mutations/assign_staff.rb
Comment (gigxz) [inline drivers/hmis/app/graphql/mutations/assign_staff.rb]: Are these expected to be "fixable" validation errors? If not, we should raise instead

File: drivers/hmis/app/graphql/mutations/unassign_staff.rb
Comment (gigxz) [inline drivers/hmis/app/graphql/mutations/unassign_staff.rb]: I don't think this is giving us much, imo clearer just to write out the `destroy!`

File: drivers/hmis/app/models/hmis/staff_assignment.rb
Comment (martha) [inline drivers/hmis/app/models/hmis/staff_assignment.rb]: If the user tries to submit a non-unique combination of [household, user, assignment type], the error message they see is "Household has already been taken." This isn't the most intuitive error message - the problem is not specifically with household, but with the combo of the 3 fields. There's probably a better railsy way to do this, maybe a way to apply a validation that is not tied to a specific field? To discuss/investigate docs more

File: drivers/hmis/app/graphql/types/hmis_schema/household.rb
Comment (martha) [inline drivers/hmis/app/graphql/types/hmis_schema/household.rb:53]: I haven't validated any permissions here. Is it safe to assume that anyone who has view permission on the household is fine to view the caseload info about that household, or should this be more restricted?

File: db/warehouse/migrate/20240717205642_add_staff_assignment_tables.rb
Comment (martha) [inline db/warehouse/migrate/20240717205642_add_staff_assignment_tables.rb]: I'll remove this comment before merging - let's discuss

File: drivers/hmis/app/graphql/types/forms/pick_list_option.rb
Comment (martha) [reply drivers/hmis/app/graphql/types/forms/pick_list_option.rb]: This pick list type should accept project as an argument, and the returned list should only include users with access to edit this enrollment.

This means we can get rid of the with_staff_assignments_enabled scope

File: drivers/hmis/app/graphql/types/hmis_schema/project.rb
Comment (martha) [inline drivers/hmis/app/graphql/types/hmis_schema/project.rb]: We don't really need the data loader here, since this isn't called in batch. Call the class method and add a comment that indicates it shouldn't be used in batch

File: drivers/hmis/app/models/hmis/hud/project.rb
Comment (martha) [inline drivers/hmis/app/models/hmis/hud/project.rb]: This should use `detect_best_config_for_project`

File: drivers/hmis/app/graphql/types/hmis_schema/household.rb
Comment (martha) [reply drivers/hmis/app/graphql/types/hmis_schema/household.rb:53]: discussed, fine to leave as is

Comment (martha) [reply drivers/hmis/app/graphql/types/hmis_schema/household.rb:53]: add pagination (assume users will add more data than we expect)
for the frontend card of current assignees, we don't have a design for pagination, so just limit the query to 100

File: drivers/hmis/app/models/hmis/staff_assignment.rb
Comment (martha) [reply drivers/hmis/app/models/hmis/staff_assignment.rb]: remove this validation (it's already checked on the db), instead add a check to the mutation where we can customize the error message

File: drivers/hmis/app/graphql/mutations/assign_staff.rb
Comment (martha) [reply drivers/hmis/app/graphql/mutations/assign_staff.rb]: Once we move the uniqueness check in here (above), we should just get rid of the .valid? check, since there is no other user-resolvable error possible

**Comment by martha** [comment]
@ttoomey, I would still like to put some time into [this todo](https://github.com/greenriver/hmis-warehouse/pull/4564/files#diff-46d7e0c7dbbcaec4b6c665a361a4af72207a48232a9aa1e844cb98995f852202R93), but otherwise would be very grateful for your review! thanks!

**Comment by ttoomey** [comment]
> @ttoomey, I would still like to put some time into [this todo] but otherwise would be very grateful for your review! thanks!

Sorry I missed this PR request, thanks for the nudge. I'll get it reviewed asap!

File: drivers/hmis/app/graphql/mutations/assign_staff.rb
Comment (ttoomey) [inline drivers/hmis/app/graphql/mutations/assign_staff.rb]: Since we only load a single record, I think the preload isn't needed. Also probably should use the viewable_by scope so we never even load a record the user can't see
```suggestion
      household = Hmis::Hud::Household.viewable_by(current_user).find_by(household_id: household_id, data_source_id: current_user.hmis_data_source_id)
```

File: drivers/hmis/app/graphql/mutations/assign_staff.rb
Comment (ttoomey) [inline drivers/hmis/app/graphql/mutations/assign_staff.rb]: This is a small thing but enclosing the check in a lock forces duplicate requests to run in series so the the behavior is consistent (avoids race condition). 

Also, it's a little more concise to call `create!` rather than new, save!

```suggestion
      # Locking on assignment_type looks odd but we can't use user (different db) and household is a view
      assignment_type.with_lock do
        errors.add(:user_id, :invalid, message: "#{user.name} is already assigned as #{assignment_type.name} for this household") if existing.exists?
        return { errors: errors } if errors.any?

        assignment = Hmis::StaffAssignment.create!(
          household: household,
          user: user,
          staff_assignment_type: assignment_type,
        )
      end
```

File: drivers/hmis/app/graphql/mutations/assign_staff.rb
Comment (ttoomey) [inline drivers/hmis/app/graphql/mutations/assign_staff.rb]: It seems like it should mirror what we have in the picklist?
```suggestion
      user = Hmis::User.can_edit_enrollments_for(household.project).find(user_id)
```

File: drivers/hmis/app/graphql/mutations/assign_staff.rb
Comment (ttoomey) [inline drivers/hmis/app/graphql/mutations/assign_staff.rb]: Perhaps this should be an input type? Input types tend to be easier to work with imo but perhaps this mutation is so simple it's not needed
```ruby
    argument :input, Types::HmisSchema::AssignStaffInput, required: true
```

File: db/warehouse/migrate/20240717205642_add_staff_assignment_tables.rb
Comment (ttoomey) [inline db/warehouse/migrate/20240717205642_add_staff_assignment_tables.rb]: It seems like the two queries where we might want indexes for are user.staff_assignments and household.staff_assignments.

Postgres can only use a composite index from left to right, so it can't be used for the household query

Perhaps this we could reorder it and then add another index for the user_id:
```suggestion
    add_index :hmis_staff_assignments, [:data_source_id, :household_id, :user_id, :hmis_staff_assignment_type_id], unique: true, where: 'deleted_at IS NULL', name: :uidx_hmis_staff_assignments
    add_index :hmis_staff_assignments, :user_id
```

File: drivers/hmis/app/graphql/types/forms/pick_list_option.rb
Comment (ttoomey) [inline drivers/hmis/app/graphql/types/forms/pick_list_option.rb]: if this is display order for the ui, perhaps order by last, first? or fullname?
`      Hmis::User.can_edit_enrollments_for(project).order(:last_name, :first_name,:id).map do |user|`

**Comment by martha** [comment]
No rush @ttoomey I just requested your review yesterday -- it's been in draft mode! :)

File: drivers/hmis/app/graphql/mutations/unassign_staff.rb
Comment (ttoomey) [inline drivers/hmis/app/graphql/mutations/unassign_staff.rb]: This is fine; looks like we have the convention of raising a runtime error in this case. Some day we should improve this though, probably should just use scope.find() which raises ActiveRecord::RecordNotFound. Then handle that error generically so we get consistent behavior across the schema. Something for future housekeeping 🧹

File: drivers/hmis/spec/factories/hmis/hud/staff_assignments.rb
Comment (ttoomey) [inline drivers/hmis/spec/factories/hmis/hud/staff_assignments.rb]: perhaps this should be a sequence so the factory can make multiple instances without violating the uniqueness constraint on name
```suggestion
    sequence(:name) { |n| "staff role #{n}" }
````

File: drivers/hmis/app/graphql/types/hmis_schema/household.rb
Comment (ttoomey) [inline drivers/hmis/app/graphql/types/hmis_schema/household.rb]: small suggestion to use id rather than deleted_at as a tie breaker
```suggestion
      scope = load_ar_association(object, :staff_assignments).order(created_at: :desc, id: desc)
      if is_currently_assigned
        scope
      else
        scope.with_deleted.where.not(deleted_at: nil)
```

File: drivers/hmis/app/graphql/types/hmis_schema/staff_assignment.rb
Comment (ttoomey) [inline drivers/hmis/app/graphql/types/hmis_schema/staff_assignment.rb]: Maybe this should be called something like "staff_assignment_name" or maybe just "role_name". In the context of a resolver, "type" suggests to me that this could be an enum

File: db/warehouse/migrate/20240717205642_add_staff_assignment_tables.rb
Comment (ttoomey) [inline db/warehouse/migrate/20240717205642_add_staff_assignment_tables.rb]: I wonder if this would be better named "hmis_staff_assignment_roles" to match the inline comment. Also "type" is a somewhat overloaded term.

File: drivers/hmis/app/graphql/mutations/unassign_staff.rb
Comment (ttoomey) [inline drivers/hmis/app/graphql/mutations/unassign_staff.rb]: I think it might be better to standardize on the `current_permission?` helper

```suggestion
      access_denied! unless current_permission?(permission: :can_edit_enrollments, entity: record.household.project)
```

File: drivers/hmis/app/models/hmis/user.rb
Comment (ttoomey) [inline drivers/hmis/app/models/hmis/user.rb]: it might be worth asking Elliot about this, we could well have something already implemented on the warehouse side, which uses similar models

**Review by ttoomey — APPROVED** [review]
This is looking great overall, thanks Martha! 

I left a bunch of suggestions that are all optional.

File: db/warehouse/migrate/20240717205642_add_staff_assignment_tables.rb
Comment (martha) [reply db/warehouse/migrate/20240717205642_add_staff_assignment_tables.rb]: thanks @ttoomey this is great 🎉

File: drivers/hmis/app/graphql/mutations/unassign_staff.rb
Comment (martha) [reply drivers/hmis/app/graphql/mutations/unassign_staff.rb]: Oh true, I'm happy to simplify it to use `find` now. And later we can handle that error generically, as you said.

File: drivers/hmis/spec/factories/hmis/hud/staff_assignments.rb
Comment (martha) [reply drivers/hmis/spec/factories/hmis/hud/staff_assignments.rb]: Great 👍

File: drivers/hmis/app/graphql/types/hmis_schema/staff_assignment.rb
Comment (martha) [reply drivers/hmis/app/graphql/types/hmis_schema/staff_assignment.rb]: I agree with you about "type" implying an enum, but I also remember Gig had concerns about "role" being overloaded too because it sounds like a permissions thing. "staff_assignment_title" (like job title?) is the only other good alternative that I have been able to come up with so far, but I feel hesitant to change things around in the spec that we agreed on before Gig left

File: drivers/hmis/app/models/hmis/user.rb
Comment (martha) [reply drivers/hmis/app/models/hmis/user.rb]: @eanders , do you know if we do anything similar to this in the warehouse? From what I could find, not yet:
https://github.com/greenriver/hmis-warehouse/blob/9755f5b77568378ef4f160713db8579848c4d63d/app/models/user.rb#L73-L78
is the most similar, but doesn't include a scope that checks for all users with permission *for* a particular entity.

Comment (eanders) [reply drivers/hmis/app/models/hmis/user.rb]: We usually do the check on the entity `scope :viewable_by, -> (user)`.  (See `app/models/grda_warehouse/hud/project.rb#viewable_by` for one example.)

The trick with going the other way is that you have to walk the inheritance chain, users can inherit access to a project from the project's organization for instance.

I think this would be something like (untested code):
```
scope "#{permission}_for", ->(entity) do
 user_ids = Hmis::AccessControl.joins(:role, :access_group, user_group: :users).
    preload(user_group: :user_group_members).
    merge(Hmis::Role.where(permission => true)).
    merge(Hmis::AccessGroup.contains(entity)). # I think this `contains` scope isn't complete since it doesn't walk the inheritance tree (but maybe it should?)
    select(Hmis::User.arel_table[:id]) # selecting the ids so we can ensure the returned scope doesn't include any of the complexity 
  User.where(id: user_ids)
end

```

Looking through the code, I think it would be helpful to add a `contains_with_inherited` scope on `Hmis::AccessGroup`.  That will be a bit of a pain to write, but should make the above useful.

Comment (martha) [reply drivers/hmis/app/models/hmis/user.rb]: Thanks @eanders this is extremely helpful. Does this look like I am on the right track?
https://github.com/greenriver/hmis-warehouse/pull/4564/files#diff-c4a75867ebfe78fba438a93af5a60cddcd44463aaa2be29eac798cc3902d30a0R49-R62

- Do you think it is OK to leave the other entity types as 'not implemented' for now until a use case arises?

Comment (eanders) [reply drivers/hmis/app/models/hmis/user.rb]: Hmm, so I think I must not have been very clear.  (Happy to huddle about it if that'd be easier.)  What you have still doesn't include the inheritance (as far as I can see.)  When checking for projects, you need to include the organizations and data sources that also contain that project since permission may be granted further up the chain.  

I'll drop a comment on that scope to see if I can get closer.

File: drivers/hmis/app/models/hmis/access_group.rb
Comment (eanders) [inline drivers/hmis/app/models/hmis/access_group.rb]:
```suggestion
  additional_entities = case entity
  when Hmis::Hud::Project
    {
      'Hmis::Hud::Organization' => entity.organization.id, 
      'GrdaWarehouse::DataSource' => entity.data_source.id,
    } 
  when Hmis::Hud::Organization
    {
      'GrdaWarehouse::DataSource' => entity.data_source.id,
    }
  when Hmis::DataSource
    {}
  else
    raise "Unsupported entity type: #{entity.class.name}"
  end
  a_t = Hmis::GroupViewableEntity.arel_table
  # Find the collections involving this entity directly
  query = a_t[:entity_type].eq(entity.class.name).and(a_t[:entity_id].eq(entity.id))
  # Add collections where this entity is a sub-entity of the assigned entity
  additional_entities.each do |klass, id|
    query = query.or(a_t[:entity_type].eq(klass).and(a_t[:entity_id].eq(id)))
  end
  # Fetch the collections that include this entity, either directly, or through inehritance
  collection_ids = Hmis::GroupViewableEntity.where(query).pluck(:collection_id)
  # Return a clean AccessGroup scope
  where(id: collection_ids)
```

Comment (martha) [reply drivers/hmis/app/models/hmis/access_group.rb]: Thanks @eanders this makes sense and is very thorough, I'm happy to commit it, but I think maybe some of this work of finding the entities through inheritance is being done already through the `group_viewable_entity_projects` view? I think that's why the code in [ProjectAccessLoader](https://github.com/greenriver/hmis-warehouse/blob/9755f5b77568378ef4f160713db8579848c4d63d/drivers/hmis/app/models/hmis/hud/project_access_loader.rb#L16-L20) works, which is what my sketch was based on, even though it does not explicitly make queries for data source and organization based collections.

Comment (eanders) [reply drivers/hmis/app/models/hmis/access_group.rb]: Oh!  Sorry, I'm not that familiar with the HMIS permission scheme, it was, at one point, very similar to the warehouse version, but seems to have matured.  Yes, `hmis_group_viewable_entity_projects` looks to include the same relationships I added.  Do you (or maybe @ttoomey) know how to take advantage of that table?

It looks like `Hmis::Hud::Project.joins(:group_viewable_entity_projects)` might be the ticket, but I'm definitely not following that completely.  Maybe you had the project version covered already.  The Data Sources are simple, but maybe we should build a new view for organizations?  (Or maybe we only need to determine which projects someone has access to, which might mean what you had was sufficient.)

Comment (martha) [reply drivers/hmis/app/models/hmis/access_group.rb]: Hm, ok, yes I think you're right that `.joins(:group_viewable_entity_projects)` is the ticket, because of how this relationship is set up on the [Project model](https://github.com/greenriver/hmis-warehouse/blob/9755f5b77568378ef4f160713db8579848c4d63d/drivers/hmis/app/models/hmis/hud/project.rb#L46) with `has_many :group_viewable_entities, through: :group_viewable_entity_projects`.

I also discovered that it looks like there are already helpers for `includes_entities` with inheritance on the [GroupViewableEntity](https://github.com/greenriver/hmis-warehouse/blob/9755f5b77568378ef4f160713db8579848c4d63d/drivers/hmis/app/models/hmis/group_viewable_entity.rb#L53-L55), though they seemed to be a bit out-of-date and untested, so I fixed one or two bugs and added some spec tests to reassure us of the behavior.

Comment (martha) [reply drivers/hmis/app/models/hmis/access_group.rb]: @ttoomey and/or @eanders (no rush) I've re-requested your review for this part specifically -- I'd be very glad for either one of you to give it a second pair of eyes before I merge, since I haven't done that much work on permissions yet.

File: db/warehouse/migrate/20240717205642_add_staff_assignment_tables.rb
Comment (eanders) [reply db/warehouse/migrate/20240717205642_add_staff_assignment_tables.rb]: @martha are these actually "relationships"?  I feel like both `roles` and `types` are heavily used, but if this is describing the relationship between the client and the staff member, something like **Case Manager** or **Housing Navigator**, I'd maybe go with `hmis_staff_assignment_relationships`, though `hmis_staff_assignment_relationship_types` also feels right to me...  Maybe just `hmis_staff_relationships`.

File: drivers/hmis/app/graphql/types/hmis_schema/staff_assignment.rb
Comment (eanders) [reply drivers/hmis/app/graphql/types/hmis_schema/staff_assignment.rb]: Feel free to ignore my suggestion about `relationship` above (and here), but that's the other wording that feels vaguely right to me.

**Review by eanders — APPROVED** [review]
Looks good to me.  Thanks @martha !

File: drivers/hmis/app/graphql/types/hmis_schema/staff_assignment.rb
Comment (martha) [reply drivers/hmis/app/graphql/types/hmis_schema/staff_assignment.rb]: Thanks, hm I like `relationship` too! I'll go ahead with the change

**Pull request merged**
